### PR TITLE
Add support for innr SP 234 smart plug

### PR DESCRIPTION
**Pull request by SwoopX**
See #5256.

### CODE REVIEW

**Comment by xconverge** [comment]
do we need to update de_web_plugin_private.h

```
        case VENDOR_INNR:
            return prefix == jennicMacPrefix ||
                   prefix == silabs4MacPrefix ||
                   prefix == silabs10MacPrefix;
```

**Comment by SwoopX** [comment]
No, not really. But I forgot something else...

**Comment by xconverge** [comment]
15:47:30:952 remove binding from 0x847127FFFE9805D2 cluster 0x2000 due non existing cluster

**Comment by xconverge** [comment]
I cant find anything related to cluster 0x2000 anywhere....?

**Comment by xconverge** [comment]
I am not seeing anything about energy with this running:

```
 "6":{
      "etag":"a2610b7ce85fdc5213648d08b7d00fb2",
      "hascolor":false,
      "lastannounced":"2021-08-29T01:14:19Z",
      "lastseen":"2021-08-29T01:17Z",
      "manufacturername":"innr",
      "modelid":"SP 234",
      "name":"On/Off plug-in unit 6",
      "state":{
         "alert":"none",
         "on":true,
         "reachable":true
      },
      "swversion":"2.8",
      "type":"On/Off plug-in unit",
      "uniqueid":"84:71:27:ff:fe:98:05:d2-01"
   }
```

**Comment by xconverge** [comment]
Disregard, I redeployed, and had to delete/repair the device and it looks good now!

image
image

File: simple_metering.cpp [hunk 1]
```diff
@@ -93,6 +93,7 @@ void DeRestPluginPrivate::handleSimpleMeteringClusterIndication(const deCONZ::Ap
                     consumption = static_cast<quint64>(round((double)consumption / 10.0)); // 0.1 Wh -> Wh
                 }
                 else if (modelId == QLatin1String("SP 120") ||                    // innr
+                         modelId == QLatin1String("SP 234") ||                    // innr
```
Comment (xconverge) [inline simple_metering.cpp:96]: I am not sure this is reporting correctly, what units is consumption supposed to be in, Wh?

Right now the outlet is reports consumption at 0.29 kWh and I havent really used it yet, that seems like more than I would expect for just powering the radio...?

Comment (xconverge) [reply simple_metering.cpp:96]: I will use it properly this weekend and get a better feel, I can also plug a lamp + incandescent in for an hour and get the real answer later this weekend to get a more concrete answer other than a "feeling"

**Comment by xconverge** [comment]
I have an email out to innr...

electrical measurement shows power divisor is 10, but that would make this value off by a factor of 10. For now I would hold off on merging this until I get feedback from innr

image

image

**Comment by manup** [comment]
A DDF would be very welcome, as a reference you might have a look at the SP120 DDF:

https://github.com/dresden-elektronik/deconz-rest-plugin/blob/master/devices/innr/sp_120.json

**Comment by xconverge** [comment]
@manup I wasn't able to get it working 2 weeks ago so was going to give it a new try when a new beta was released, I can give it a try again though I was unable to get any data out

**Comment by xconverge** [comment]
Yea I tried again and cant get any data besides on/off to come out or showup in the restapi. Always null

```
    "22": {
      "config": {
        "on": true,
        "reachable": true
      },
      "etag": "4a3a3b839e4e206b75c301a86cbb0554",
      "lastannounced": "2021-11-07T22:24:16Z",
      "lastseen": "2021-11-08T22:30Z",
      "manufacturername": "innr",
      "modelid": "SP 234",
      "name": "Power 22",
      "state": {
        "current": null,
        "lastupdated": "none",
        "power": null,
        "voltage": null
      },
      "type": "ZHAPower",
      "uniqueid": "84:71:27:ff:fe:98:05:d2-01-0b04"
    },
```

**Comment by manup** [comment]
Can you please share the DDF file, I can have a look and compare with the screenshots from the issue.

Is the *deCONZ > Panels > Control > Normal* mode activated?
Currently the mode is reset after each restart of deCONZ.

**Comment by xconverge** [comment]
I had no idea that existed, that certainly helps. I got power consumption and power out now, my ddf is literally the sp120 one with model and product changed to 234

I will need to play with this a bit more to figure out why there are so many duplicate entries and stuff. Are there supposed to be double power and double consumption entries as shown in my message above?

And then when the outlet is off it shows there being no sensors?

```
15:42:49:479 GW firmware version is up to date: 0x26700700
15:43:03:151 ZCL attribute report 0x5C0272FFFE67C7D5 for cluster: 0x0008, ep: 0x01, frame control: 0x08, mfcode: 0x0000 
15:43:06:979 ZCL read attr 0x847127FFFE9805D2, ep: 0x01, cl: 0x0B04, attr: 0x0508, mfcode: 0x0000, aps.id: 169, zcl.seq: 253
15:43:07:014 No power sensor found for 0x847127FFFE9805D2, endpoint: 0x01
15:43:07:017 ZCL read attr 0x847127FFFE9805D2, ep: 0x01, cl: 0x0B04, attr: 0x050B, mfcode: 0x0000, aps.id: 172, zcl.seq: 254
15:43:07:063 No power sensor found for 0x847127FFFE9805D2, endpoint: 0x01
15:43:07:066 ZCL read attr 0x847127FFFE9805D2, ep: 0x01, cl: 0x0B04, attr: 0x0505, mfcode: 0x0000, aps.id: 174, zcl.seq: 255
15:43:07:108 No power sensor found for 0x847127FFFE9805D2, endpoint: 0x01
15:43:10:597 ZCL attribute report 0x5C0272FFFE67C7D5 for cluster: 0x0006, ep: 0x01, frame control: 0x08, mfcode: 0x0000 
15:43:18:979 ZCL read attr 0x847127FFFE9805D2, ep: 0x01, cl: 0x0702, attr: 0x0000, mfcode: 0x0000, aps.id: 243, zcl.seq: 0
15:43:19:011 No consumption sensor found for 0x847127FFFE9805D2, endpoint: 0x01
```

**Comment by xconverge** [comment]
This device has a wrong value for all current FW as mentioned above, so I think I need to make this adjustment in my DDF for power calculation, but its going to be kind of a mess because they told me they are fixing it for new batches...

```
    "21": {
      "config": {
        "on": true,
        "reachable": true
      },
      "etag": "dbc3c33a41975718b9cc1e8f868f069e",
      "lastannounced": "2021-11-07T22:24:16Z",
      "lastseen": "2021-11-08T23:34Z",
      "manufacturername": "innr",
      "modelid": "SP 234",
      "name": "Consumption 21",
      "state": {
        "consumption": 28,
        "lastupdated": "none"
      },
      "swversion": "2.8",
      "type": "ZHAConsumption",
      "uniqueid": "84:71:27:ff:fe:98:05:d2-01-0702"
    },
    "22": {
      "config": {
        "on": true,
        "reachable": true
      },
      "etag": "7839de9d739118e91efc97ae1674711b",
      "lastannounced": "2021-11-07T22:24:16Z",
      "lastseen": "2021-11-08T23:34Z",
      "manufacturername": "innr",
      "modelid": "SP 234",
      "name": "Power 22",
      "state": {
        "current": 104,
        "lastupdated": "none",
        "power": 6,
        "voltage": 121
      },
      "swversion": "2.8",
      "type": "ZHAPower",
      "uniqueid": "84:71:27:ff:fe:98:05:d2-01-0b04"
    },
```

**Comment by manup** [comment]
I think in next releases we need to add the possibility to select a DDF file not only by manufacturer name and model id but also optionally with other characteristics like the firmware version.

As a workaround the "eval" expression has also access to other attributes, for example:

```
Item.val = R.item('attr/swversion').val === '2.8' ? Attr.val / 123 : Attr.val / 456;
```

So here depending on the firmware version either `Attr.val / 123`or `Attr.val / 456` will be calculated.
Relative ugly but should work ;)

**Comment by xconverge** [comment]
@manup 

I have a .pcap capture of the dropoff (happens with both devices) if anyone wants to take a peek. I will be sending it to innr also.

Dropoff of device 84:71:27:FF:FE:98:06:53 happens at ~15:28:23:141 which is ~200 seconds in the log

Network key for decrypting the traffic is 1fe36ca7db7bb8a4d3e110e3cb6868b6

[dropoff.zip](https://github.com/dresden-elektronik/deconz-rest-plugin/files/7530985/dropoff.zip)

**Comment by SwoopX** [comment]
Just a theory on this one here.

Based on the sniff, it looks like the device is desperately seeking some network address response, which it will never get due to a bug in 2.13.1 where the extended address is set incorrectly (seen on the trailing `:00:00:00`). The keep-aive got interestingly answered in between, but it only uses the short address. Too bad that the IEEE level is heavily underrepresented...

So maybe, its passing out due to the missing responses. That one got fixed in 2.13.2, but requires the device to be re-paired in order to get a correct address.

grafik

So you might want to try out the latest beta and keep an eye on the behavior once more. Crossing fingers it does the trick.

As a side note to innr: great that you shared the response of them and that they acknowledged the mistake. On the other hand, I cannot really understand they've provided you with the firmware update to get it fixed. That will most likely also bring other projects kinda out of sync.

**Comment by xconverge** [comment]
Reply from innr:
"My colleague had a quick look and think it has to do with the Keep-Alive cluster and what you can try is to disable support for the Keep-Alive cluster (cluster ID 0x0025) on your bridge. Unless you need a Zigbee Smart Energy network with support for TC Swap-out, the Keep-Alive cluster is not needed.
 
He has also picked this up with our other developers, and I will update you as always when I have new information. Hopefully this workaround works for now."

I haven't tested 2.13.02 yet due to deployment issues I am facing but will as soon as possible

**Comment by xconverge** [comment]
@SwoopX unfortunately the problem still exists with 2.13.02

**Comment by xconverge** [comment]
@SwoopX you can close this PR, Copy pasting the sp120 ddf  and making changes works now

**Comment by manup** [comment]
Closing since replaced by PR https://github.com/dresden-elektronik/deconz-rest-plugin/pull/6513

**Pull request closed without merging**